### PR TITLE
Enable version negotiation with Docker client

### PR DESCRIPTION
**Pull request by pkwarren**
Update Docker client to negotiate the client version and also fix typo in error handling.

### CODE REVIEW

File: private/buf/cmd/buf/command/alpha/plugin/pluginpush/pluginpush.go [hunk 1]
```diff
@@ -199,7 +199,7 @@ func run(
 	if flags.Image != "" {
 		inspectResponse, err := client.Inspect(ctx, flags.Image)
 		if err != nil {
-			return nil
+			return err
```
Comment (pkwarren) [inline private/buf/cmd/buf/command/alpha/plugin/pluginpush/pluginpush.go:202]: 🙄

File: private/bufpkg/bufplugin/bufplugindocker/docker.go [hunk 2]
```diff
@@ -225,6 +225,11 @@ func NewClient(logger *zap.Logger, cliVersion string, options ...ClientOption) (
 		client.WithHTTPHeaders(map[string]string{
 			"User-Agent": BufUpstreamClientUserAgentPrefix + cliVersion,
 		}),
+		// NOTE: This setting is not safe when a client is called from multiple goroutines:
```
Comment (bufdev) [inline private/bufpkg/bufplugin/bufplugindocker/docker.go:228]: This feels like a massive potential bug in the future - are we sure we want to do this?

File: private/bufpkg/bufplugin/bufplugindocker/docker.go [hunk 3]
```diff
@@ -225,6 +225,11 @@ func NewClient(logger *zap.Logger, cliVersion string, options ...ClientOption) (
...
+		// - https://github.com/moby/moby/issues/43729#issuecomment-1174446640
+		//
+		// Our usage is for the CLI with a single goroutine so we're ok to enable for compatibility.
+		client.WithAPIVersionNegotiation(),
```
Comment (mfridman) [inline private/bufpkg/bufplugin/bufplugindocker/docker.go:232]: Are we sure we want to bake this into the CLI?

**Review by mfridman — APPROVED** [review]

File: private/bufpkg/bufplugin/bufplugindocker/docker.go (hunk 2 above)
Comment (rhbuf) [reply private/bufpkg/bufplugin/bufplugindocker/docker.go:228]: right now it just breaks when used against any server that is a lower version, not sure what our other mitigations are, the docker team [are aware of this](https://github.com/moby/moby/issues/43729#issuecomment-1174446640) so will hopefully fix in the near future
```
Error response from daemon: client version 1.42 is too new. Maximum supported API version is 1.41
```

Comment (pkwarren) [reply private/bufpkg/bufplugin/bufplugindocker/docker.go:228]: We are only using this to push custom plugins - there are no other code paths that use these Docker abstractions and no concurrency in play there.

The alternative is that we hardcode an earlier version of the Docker API version that is compatible with most systems.

File: private/bufpkg/bufplugin/bufplugindocker/docker.go (hunk 3 above)
Comment (rhbuf) [reply private/bufpkg/bufplugin/bufplugindocker/docker.go:232]: unfortunately it cannot be set from the [env](https://pkg.go.dev/github.com/docker/docker/client#FromEnv) like most other options

Comment (mfridman) [reply private/bufpkg/bufplugin/bufplugindocker/docker.go:232]: i guess you could set it manually `DOCKER_API_VERSION=1.x`, but that's obviously not great (nor recommended).

this seems reasonable, I just hope the negotiation isn't buggy.

File: private/bufpkg/bufplugin/bufplugindocker/docker.go (hunk 2 above)
Comment (pkwarren) [reply private/bufpkg/bufplugin/bufplugindocker/docker.go:228]: PTAL - I updated to force version negotiation when creating the client so reads/writes to those fields should no longer race.

File: private/bufpkg/bufplugin/bufplugindocker/docker.go (hunk 3 above)
Comment (pkwarren) [reply private/bufpkg/bufplugin/bufplugindocker/docker.go]: The updated code will skip setting the version if you've provided an override via the env var:

https://github.com/moby/moby/blob/d7573ab8672555762688f4c7ab8cc69ae8ec1a47/client/client.go#L227-L244

**Review by bufdev — APPROVED** [review]

**Pull request merged**